### PR TITLE
Decouple S3 object actions, add unit tests

### PR DESCRIPTION
**Pull request by abrooksv**
* Decouple S3 object actions from the TreeTable and Project arguments
* Add unit tests

## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/editor/S3EditorDataKeys.kt [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+    /**
+     * Returns the Bucket for the viewer
+     */
+    val BUCKET = DataKey.create<S3VirtualBucket>("aws.s3.bucketViewer.bucket")
+
+    /**
+     * Returns the S3 bucket viewer table
+     */
+    val BUCKET_VIEWER = DataKey.create<S3TreeTable>("aws.s3.bucketViewer")
```
Comment (hunterwerlla) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/editor/S3EditorDataKeys.kt:22]: `BUCKET_TABLE`? viewer is a confusing name IMO

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/editor/S3TreeTable.kt [hunk 2]
```diff
@@ -233,6 +234,13 @@ class S3TreeTable(
         }
     }
 
+    override fun getData(dataId: String): Any? = when {
+        S3EditorDataKeys.SELECTED_NODES.`is`(dataId) -> this.getSelectedNodes().filterNot { it is S3TreeContinuationNode<*> }
+        S3EditorDataKeys.BUCKET.`is`(dataId) -> this.bucket
+        S3EditorDataKeys.BUCKET_VIEWER.`is`(dataId) -> this
+        else -> null
```
Comment (hunterwerlla) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/editor/S3TreeTable.kt:241]: throw instead of null?

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/objectActions/ViewObjectVersionAction.kt [hunk 3]
```diff
@@ -3,18 +3,20 @@
...
-
-    override fun performAction(node: S3TreeNode) {
+class ViewObjectVersionAction : SingleS3ObjectAction(message("s3.version.history.view"), AllIcons.Actions.ShowAsTree) {
+    override fun performAction(dataContext: DataContext, node: S3TreeNode) {
         if (node is S3TreeObjectNode) {
             node.showHistory = true
-            treeTable.refresh()
+
+            // TODO: Can we expand the node too
+            dataContext.getRequiredData(S3EditorDataKeys.BUCKET_VIEWER).refresh()
```
Comment (hunterwerlla) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/objectActions/ViewObjectVersionAction.kt:19]: Doesn't this already expand the node on refresh? Does this mean something different?

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/editor/S3EditorDataKeys.kt (hunk 1 above)
Comment (abrooksv) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/editor/S3EditorDataKeys.kt:22]: Agree :D

File: core/tst/software/aws/toolkits/core/utils/DelegateSdkConsumers.kt [hunk 4]
```diff
@@ -41,9 +41,10 @@ class DelegateSdkConsumers(private val sdkClass: Class<*>) : Answer<Any> {
     }
 }
 
-inline fun <reified T : SdkClient> delegateMock(): T = Mockito.mock(
+inline fun <reified T : SdkClient> delegateMock(verboseLogging: Boolean = false): T = Mockito.mock(
```
[the remark this answers is not in the document]
Comment (abrooksv) [reply core/tst/software/aws/toolkits/core/utils/DelegateSdkConsumers.kt:44]: it's not, I used it when making the mocks

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/editor/S3TreeTable.kt (hunk 2 above)
Comment (abrooksv) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/editor/S3TreeTable.kt:241]: Nope, API contract says return null

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/objectActions/ViewObjectVersionAction.kt (hunk 3 above)
Comment (abrooksv) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/objectActions/ViewObjectVersionAction.kt:19]: Nope, we don't expand versions when we load them today. User has to go in and expand the node

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/editor/S3TreeTable.kt (hunk 2 above)
Comment (hunterwerlla) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/editor/S3TreeTable.kt:241]: Ah, sure. It's kind of weird though because that is something we wouldn't want to silently fail if we messed up

Comment (abrooksv) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/editor/S3TreeTable.kt:241]: DataProvider traverses up the UI tree, so null means ask the parent.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/objectActions/CopyPathAction.kt [hunk 5]
```diff
@@ -3,22 +3,21 @@
...
-class CopyPathAction(private val project: Project, treeTable: S3TreeTable) : SingleS3ObjectAction(treeTable, message("s3.copy.path")) {
-    override fun performAction(node: S3TreeNode) {
+class CopyPathAction : SingleS3ObjectAction(message("s3.copy.path")) {
+    override fun performAction(dataContext: DataContext, node: S3TreeNode) {
         CopyPasteManager.getInstance().setContents(StringSelection(node.key))
-        S3Telemetry.copyPath(project)
+        S3Telemetry.copyPath(dataContext.getRequiredData(CommonDataKeys.PROJECT))
     }
 
-    override fun enabled(node: S3TreeNode): Boolean = node !is S3TreeObjectVersionNode && node !is S3TreeContinuationNode<*> && node !is S3TreeErrorNode
```
Comment (hunterwerlla) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/objectActions/CopyPathAction.kt:23]: why was `&& node !is S3TreeContinuationNode<*> && node !is S3TreeErrorNode` removed?

**Review by hunterwerlla — COMMENTED** [review]
UI tests failed

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/objectActions/CopyPathAction.kt (hunk 5 above)
Comment (abrooksv) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/objectActions/CopyPathAction.kt:23]: DataContext filters out those nodes

https://github.com/aws/aws-toolkit-jetbrains/pull/2401/files#diff-b26b311a1cfe610f3d4b3dea94c8e83935031bd9dce7a6479963c123e0f98637R65

Comment (abrooksv) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/objectActions/CopyPathAction.kt:23]: But maybe we shouldn't? So that selecting one of those and a valid node will disable these actions?

Comment (rli) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/objectActions/CopyPathAction.kt:23]: can we have tests on that data provider?

Comment (abrooksv) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/objectActions/CopyPathAction.kt:23]: Ya, we should get those added

Comment (rli) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/objectActions/CopyPathAction.kt:23]: > But maybe we shouldn't? So that selecting one of those and a valid node will disable these actions?

how about just making errors/continuation nodes nonselectable?

Comment (abrooksv) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/objectActions/CopyPathAction.kt:23]: I went looking for a Swing API for that and did not find one. Do you know of a way?

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/editor/S3TreeTable.kt [hunk 6]
```diff
@@ -158,8 +158,10 @@ class S3TreeTable(
     }
 
     init {
-        // Associate the drop target listener with this instance which will allow uploading by drag and drop
-        DropTarget(this, dropTargetListener)
+        if (!ApplicationManager.getApplication().isHeadlessEnvironment) {
```
Comment (hunterwerlla) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/editor/S3TreeTable.kt:161]: Why was this added? add a comment on why?

Comment (abrooksv) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/editor/S3TreeTable.kt:161]: Sure, DND can't be started when we are running headless (aka unit tests)

Comment (abrooksv) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/editor/S3TreeTable.kt:161]: using isUnitTestMode probably clearer

**Review by hunterwerlla — APPROVED** [review]

File: jetbrains-core/tst/software/aws/toolkits/jetbrains/services/s3/objectActions/UploadObjectActionTest.kt [hunk 7]
```diff
@@ -89,6 +91,26 @@ class UploadObjectActionTest : ObjectActionTestBase() {
...
+        val dir = S3TreeDirectoryNode(s3Bucket, null, "path1/")
+        val nodes = listOf(
+            S3TreeErrorNode(s3Bucket, dir)
+        )
+
+        assertThat(sut.updateAction(nodes).isEnabled).isFalse
+    }
+
+    @Test
+    fun `uplaod action is disabled on continuation node`() {
```
Comment (rli) [inline jetbrains-core/tst/software/aws/toolkits/jetbrains/services/s3/objectActions/UploadObjectActionTest.kt:105]: "upload"

**Review by rli — APPROVED** [review]

File: jetbrains-core/tst/software/aws/toolkits/jetbrains/services/s3/objectActions/UploadObjectActionTest.kt (hunk 7 above)
Comment (abrooksv) [reply jetbrains-core/tst/software/aws/toolkits/jetbrains/services/s3/objectActions/UploadObjectActionTest.kt]: nice catch, thx
